folsom-opencensus/src/main/java/com/spotify/folsom/opencensus/OpenCensusTracer.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (c) 2019 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.spotify.folsom.opencensus;
+
+import static io.opencensus.trace.AttributeValue.stringAttributeValue;
+import static io.opencensus.trace.Span.Kind.CLIENT;
+import static java.util.Objects.requireNonNull;
+
+import com.spotify.folsom.Span;
+import com.spotify.folsom.Tracer;
+import io.opencensus.trace.Tracing;
+import java.util.concurrent.CompletionStage;
+
+class OpenCensusTracer implements Tracer {
+
+  @Override
+  public Span span(final String name, final CompletionStage<?> future, final String operation) {
+    return internalSpan(name, future, operation, null);
+  }
+
+  @Override
+  public Span span(
+      final String name,
+      final CompletionStage<?> future,
+      final String operation,
+      final String key) {
+    requireNonNull(key);
+    return internalSpan(name, future, operation, key);
+  }
+
+  @SuppressWarnings("MustBeClosedChecker")
+  private Span internalSpan(
+      final String name,
+      final CompletionStage<?> future,
+      final String operation,
+      final String key) {
+    requireNonNull(name);
+    requireNonNull(future);
+    requireNonNull(operation);
+
+    final io.opencensus.trace.Tracer tracer = Tracing.getTracer();
+
+    final io.opencensus.trace.Span ocSpan =
```

Comment:
this is the trickiest part of tracing especially with async operations happening...

with this implicit context propagation traces spans might not be linked when they go from a say an apollo thread pool -> memcached thread pool. 

if this is indeed the case here, the `span` can be added as parameter to these methods and passed in with the memcached operation. This would then become `tracer.spanBuilderWithExplicitParent()`



https://github.com/census-instrumentation/opencensus-specs/blob/master/trace/Span.md#how-span-interacts-with-context

Reply:
Yeah, this assumes that the context is propagated to the calling thread/gRPC Context. Once in Folsom-land, this should works as expected. But, this would be equally true for tracing in say Bigtable/Spanner, right (I've only looked at the Spanner client and it seems to use the same pattern as here)? If so, I believe we should make sure we propagate the tracing context correctly at least out of the box in Apollo, is this the case today?

Reply:
we should be propagating the context in apollo correctly as the traces to any gRPC based service (bigtable/pubsub/spanner) are linked already. have you tried testing this code with an apollo services yet?

Reply:
I haven't yet, but got a service hopefully lined up

Reply:
I think this pattern is generally used for many libraries, Bigtable does it similarly. It may be useful to allow passing in the parent span instead of solely relying on `getCurrentSpan` - that way if the calling code is doing something unusual there's a way to keep the context correctly linked.

Reply:
@hexedpackets do you have a reference handy to how this is done in Bigtable, would be curious as to how they opted to design it? I tried to find, but not sure I'm looking in the right place.

For now, I would not want to change the Folsom API in order to support tracing, but if we see more requests from users on setting explicit parent spans I think we should look closer at how to do this in the future.

Reply:
That makes sense, I think the use case of needing to pass in the parent span should be pretty rare.

Bigtable [creates a span when the operation is instantiated](https://github.com/googleapis/cloud-bigtable-client/blob/master/bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/AbstractRetryingOperation.java#L149) and then closes it in a callback [after the operation ends](https://github.com/googleapis/cloud-bigtable-client/blob/master/bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/AbstractRetryingOperation.java#L161). I don't this this is passed around, but various functions in the class add tags and annotations to the span when they're run.